contracts/azorius/LinearERC721Voting.sol

```diff
@@ -0,0 +1,292 @@
+
+        emit StrategySetUp(_azoriusModule, _owner);
+    }
+
+    function updateVotingPeriod(uint32 _votingPeriod) external onlyOwner {
+        _updateVotingPeriod(_votingPeriod);
+    }
+
+    function updateQuorumThreshold(uint256 _quorumThreshold) external onlyOwner {
+        _updateQuorumThreshold(_quorumThreshold);
+    }
+
+    function getProposalVotes(uint32 _proposalId) external view
+        returns (
+            uint256 noVotes,
+            uint256 yesVotes,
+            uint256 abstainVotes,
+            uint32 startBlock,
+            uint32 endBlock
+        )
+    {
+        noVotes = proposalVotes[_proposalId].noVotes;
+        yesVotes = proposalVotes[_proposalId].yesVotes;
+        abstainVotes = proposalVotes[_proposalId].abstainVotes;
+        startBlock = proposalVotes[_proposalId].votingStartBlock;
+        endBlock = proposalVotes[_proposalId].votingEndBlock;
+        // TODO why didn't we return has voted data in ERC20 strategy?
+    }
+
+    // voting requires providing the NFT addresses and ids, as IERC721 does not have a method
+    // for determining which NFT ids a particular address holds 
+    function vote(uint32 _proposalId, uint8 _support, bytes memory _nftData) external {
+        ( 
+            address[] memory tokenAddresses,
+            uint256[] memory tokenIds  
+        ) = abi.decode(_nftData, (address[], uint256[]));
+
+        uint256 weight = getWeight(_proposalId, tokenAddresses, tokenIds, msg.sender);
+
+        if (weight == 0) revert NoVotingWeight();
+
+        _vote(_proposalId, msg.sender, _support, weight);
+    }
+
+    // verifies the voter holds the NFTs and returns the total weight associated with their tokens
+    // the frontend will need to determine whether an address can vote on a proposal, as it is possible
+    // to vote twice if you get more weight later on
+    function getWeight(
+        uint256 _proposalId,
+        address[] memory _tokenAddresses,
+        uint256[] memory _tokenIds,
+        address _voter
+    ) internal returns (uint256) {
+
+        uint256 weight = 0;
+
+        for (uint256 i = 0; i < _tokenAddresses.length; i++) {
+
+            address tokenAddress = _tokenAddresses[i];
+            uint256 tokenId = _tokenIds[i];
+
+            // ensure the token hasn't voted already, and the voter actually holds the token
+            if (
+                proposalVotes[_proposalId].hasVoted[tokenAddress][tokenId] == true || 
+                _voter != IERC721(tokenAddress).ownerOf(tokenId)
+            ) {
+                continue;
+            }
+            
+            for (uint j = 0; j < governanceTokens.length; j++) {
+                GovernanceNFT memory governance = governanceTokens[j];
+                IERC721 token = governance.token;
+                if (address(token) == tokenAddress) {
+                    weight = weight + governance.weight;
+                    break;
+                }
+            }
```

Comment:
This loop could get expensive if there are lots of governance token addresses for voting configured. If the pattern I showed at the top of the contract is used, this loop shouldn't be necessary.